File: fugaku_in_practice/hands_on/sample_code/06_llio/01_localtmp/fortran/main.f90
! Copyright 2025 Research Organization for Sciecne and Techonlogy 
!-----------------------------------------------------------------

    program main

    use mpi

    implicit none

    integer :: array_size
    real(kind=8), allocatable :: A(:)

    integer :: unit_no = 20
    integer :: input_no = 10
    integer :: output_no
    character(len=132) :: argv
    character(len=132) :: input_file
    character(len=132) :: output_file

    integer :: ierr
    integer :: nprocs, myrank
    integer, parameter :: main_rank = 0

    integer :: n_loop
    integer :: ii
    integer :: i_skip
    real(kind=8) :: t1, t2
    real(kind=8) :: total1, total2
    real(kind=8) :: x


    call MPI_Init(ierr)
    call MPI_Comm_size(MPI_COMM_WORLD, nprocs, ierr)
    call MPI_Comm_rank(MPI_COMM_WORLD, myrank, ierr)

    if (command_argument_count() >= 1) then
       call get_command_argument(1, argv)
       input_file = adjustl(argv)

       open (unit=input_no, file=input_file, status='old', iostat=ierr, action='read')

       if (ierr /= 0) then
          write(0,*) 'file open error'
          call MPI_Finalize(ierr)
          stop
       endif

       read(input_no,*) array_size
       read(input_no,*) n_loop
       read(input_no,*) i_skip
       read(input_no,'(A)') output_file

       close(unit=input_no) 

    else
       array_size = 20
       n_loop = 100
       i_skip = 10
       output_file = './output/output.txt'
    endif

600 format(1x,2(A,I6,2x),(A,I10,2x),(F8.3,1x,A,2x),2(A,I6,2x))

    if (myrank == main_rank) then
       write(6,600) 'NPROCS=',nprocs, &
                    'MYRANK=',myrank, &
                    'DATA_SIZE=',array_size, dble(array_size)*8.d0/(1024.d0*1024.d0),'MiB', &
                    'NLOOP=',n_loop, &
                    'ISKIP=',i_skip
    endif

    call add_myrank_to_filename(myrank, output_file)

    !!! setting of the array A
    allocate(A(1:array_size), stat=ierr)
    if (ierr /= 0) then
       write(0,'("Allocation Error: stat = ",i2)') ierr
       call MPI_Abort(MPI_COMM_WORLD, 90, ierr)
       stop
    endif

    do ii = 1, array_size
       A(ii) = 1.0d0 - 1.0d0 / real(ii, kind=8)
    enddo

    !!! calculation and file output
    output_no = unit_no + myrank

    call MPI_Barrier(MPI_COMM_WORLD, ierr)
    total1 = MPI_Wtime()
    do ii = 1, n_loop

       if (mod(ii, i_skip) == 0) then
          call MPI_Barrier(MPI_COMM_WORLD, ierr)
          t1 = MPI_Wtime()

          open (unit=output_no, file=output_file, status='replace', action='write', iostat=ierr, &
             form='unformatted', access='stream')

          if (ierr /= 0) then
             write(0,'("File Open Error:", A)') output_file
             deallocate(A)
             call MPI_Abort(MPI_COMM_WORLD, 99, ierr)
             stop
          endif

          write(output_no) A
          close(unit=output_no)

          call MPI_Barrier(MPI_COMM_WORLD, ierr)
          t2 = MPI_Wtime()
          if (myrank == main_rank) &
             write(6,'(4x,"II=",i6,2x,"ELAPSED(SEC)=",1PE15.6)') ii, t2-t1
       endif

    enddo

    call MPI_Barrier(MPI_COMM_WORLD, ierr)
    total2 = MPI_Wtime()

    if (myrank == main_rank) then
       write(6,'("MYRANK=",i6,2x,"TOTAL ELAPSED(SEC)=",1PE15.6)') &
           myrank, total2 - total1
    endif

    deallocate(A)
    call MPI_Finalize(ierr)

contains
   subroutine add_myrank_to_filename(myrank, output_file)

   implicit none

   integer, intent(in) :: myrank
   character(len=*), intent(inout) :: output_file

   character(len=4) :: c_rank
   character(len=10) :: file_ext
   integer :: period_pos


   write(c_rank,'(i0.4)') myrank
   period_pos = index(output_file, '.', back=.true.)
   file_ext = output_file(period_pos:len_trim(output_file))

   output_file = output_file(1:period_pos-1) // '_' // c_rank // trim(file_ext)

   end subroutine add_myrank_to_filename

end program main

File: fugaku_in_practice/hands_on/sample_code/06_llio/01_localtmp/cpp/main.cpp
/* Copyright 2025 Research Organization for Science and Technology */

#include <mpi.h>
#include <iostream>
#include <fstream>
#include <cstdlib>
#include <string>
#include <cmath>

std::string add_myrank_to_filename(int, std::string);

int main(int argc, char* argv[])
{
   const int main_rank = 0;
   int myrank, nprocs;
   std::string input_file;
   std::string output_file;
   FILE *ofs;

   int array_size;
   int i_skip, n_loop;
   double x;

   double t1, total1;
   double t2, total2;
   double *A;

   std::string tmp;

   MPI_Init(&argc, &argv);
   MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
   MPI_Comm_rank(MPI_COMM_WORLD, &myrank);

   if (argc > 1) {
      input_file = argv[1];

      std::ifstream infil(input_file, std::ios::in);
      if (!infil) {
         std::cerr << "File Open Error: " << input_file << std::endl;
         MPI_Finalize();
         return EXIT_FAILURE;
      }

      getline(infil, tmp);
      array_size = stoi(tmp);

      getline(infil, tmp);
      n_loop = stoi(tmp);

      getline(infil, tmp);
      i_skip = stoi(tmp);

      getline(infil, tmp);
      output_file = tmp;

      infil.close();
   } else {
      array_size = 20;
      n_loop = 100;
      i_skip = 5;
      output_file = "./output.txt";
   }

   if (myrank == main_rank) {
      std::cout << " NPROCS = " << nprocs << " ";
      std::cout << " MYRANK = " << myrank << " ";
      std::cout << " DATA_SIZE = " << array_size << " (" << array_size * 8.0 / (1024.0 * 1024.0) << " MiB) ";
      std::cout << " NLOOP = " << n_loop << " ";
      std::cout << " ISKIP = " << i_skip << std::endl;
   }

   try {
      A = new double[array_size];
   } catch (const std::bad_alloc& e) {
      std::cerr << "Memory Allocation Error: " <<  e.what() << std::endl;
      MPI_Abort(MPI_COMM_WORLD, 90);
      return EXIT_FAILURE;
   }

   for (int ii = 0; ii < array_size; ii++){
      A[ii] = 1.0 - 1.0 / (double)(ii+1);
   }

   output_file = add_myrank_to_filename(myrank, output_file);

   MPI_Barrier(MPI_COMM_WORLD);
   total1 = MPI_Wtime();

   for (int ii = 1; ii <= n_loop; ii++) {

      if (ii%i_skip == 0) {
         MPI_Barrier(MPI_COMM_WORLD);
         t1 = MPI_Wtime();

         ofs = fopen(output_file.c_str(), "wb");

         if (!ofs) {
            std::cerr << "File Open Error: " << output_file << std::endl;
            MPI_Abort(MPI_COMM_WORLD, 99);
            return EXIT_FAILURE;
         }
         fwrite(A, sizeof(double), array_size, ofs);
         fclose(ofs);

         MPI_Barrier(MPI_COMM_WORLD);
         t2 = MPI_Wtime();

         if (myrank == main_rank) {
            std::cout << "    II= " << ii << "  ELAPSED(SEC) = " << t2 - t1 << std::endl;
         }        
      }
   }
   MPI_Barrier(MPI_COMM_WORLD);
   total2 = MPI_Wtime();

   if (myrank == main_rank) {
      std::cout<< " NPROCS = " << nprocs << "  TOTAL ELAPSED(SEC) = " << total2 - total1 << std::endl;
   }

   delete[] A;
   MPI_Finalize();

   return EXIT_SUCCESS;
}

std::string add_myrank_to_filename(int myrank, std::string filename){

   std::string s_myrank = std::to_string(myrank);
   std::string file_with_rank = filename + "_" + s_myrank;

   return file_with_rank;
}
